### PR TITLE
[FLINK-12388][docs] Update the production readiness checklist

### PR DESCRIPTION
**Pull request by sjwiesman**
## What is the purpose of the change

The production readiness checklist has grown organically over the years, and while it provides valuable information, the content does not flow cohesively as it has been worked on by a number of users.

We should improve the overall structure and readability of the checklist and also update any outdated information.

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / **docs** / JavaDocs / not documented)

### CODE REVIEW

**Comment by sjwiesman** [comment]
@flinkbot attention @fhueske

File: docs/ops/production_ready.md [hunk 1]
```diff
@@ -22,76 +22,54 @@ specific language governing permissions and limitations
...
 
 As mentioned in the documentation for [savepoints]({{ site.baseurl }}/ops/state/savepoints.html), users should set uids for
-operators. Those operator uids are important for Flink's mapping of operator states to operators which, in turn, is 
-essential for savepoints. By default operator uids are generated by traversing the JobGraph and hashing certain operator 
-properties. While this is comfortable from a user perspective, it is also very fragile, as changes to the JobGraph (e.g.
-exchanging an operator) will result in new UUIDs. To establish a stable mapping, we need stable operator uids provided 
-by the user through `setUid(String uid)`.
+each operator in their `DataStream`. 
+Uids are important for Flink's mapping of operator states to operators which, in turn, is essential for savepoints. 
+By default operator uids are generated by traversing the JobGraph and hashing certain operator properties.
```
Comment (StephanEwen) [inline docs/ops/production_ready.md]: By default **,**

File: docs/ops/production_ready.md [hunk 2]
```diff
@@ -22,76 +22,54 @@ specific language governing permissions and limitations
...
+`FsStateBackend` is very fast as each state access and update is simply passing pointers to objects on the Java heap, however, it is limited by available memory within the cluster. 
+On the other hand, `RocksDB` can scale based on available disk space and is the only state backend to support incremental snapshots. 
+However, each state access and update requires (de-)serialization and potentially reading from disk which leads to average performance that is an order of magnitude slower than the memory state backends. 
+Carefully read through the [state backend documentation]({{ site.baseurl }}/ops/state/state_backends.html) to fully understand the pros and cons of each option. 
 
-### Config JobManager High Availability(HA)
+### Configure JobManager High Availability
 
-The JobManager coordinates every Flink deployment. It is responsible for both *scheduling* and *resource management*.
+The JobManager serves as a central coordinator for each Flink deployment, being responsible for both *scheduling* and *resource management*.
```
Comment (StephanEwen) [inline docs/ops/production_ready.md]: I think this section is still a bit confusing. It makes no difference if you have a single JobManager or multiple JobManagers, as long as that single JobManager process comes back (which it does on Yarn, K8s, Mesos, etc.).

The issue is that without ZK, the JobManager process cannot recover once it comes back, and neither can the other standby JobManagers in the case of multiple JMs.

**Comment by sjwiesman** [comment]
Thanks for the review, I've updated the JobManager section to be more clear and also fixed a few other grammar issues I noticed on the second pass.

**Comment by StephanEwen** [comment]
+1

merging this...

File: docs/ops/production_ready.md [hunk 3]
```diff
@@ -22,79 +22,54 @@ specific language governing permissions and limitations
...
-value as default for this parameter. The reason behind this is that high maximum parallelism can have some impact on your
-application's performance and even state sizes, because Flink has to maintain certain metadata for its ability to rescale which
-can increase with the maximum parallelism. In general, you should choose a max parallelism that is high enough to fit your
-future needs in scalability, but keeping it as low as possible can give slightly better performance. In particular,
-a maximum parallelism higher that 128 will typically result in slightly bigger state snapshots from the keyed backends.
+### Set An Explicit Max Parallelism
 
-Notice that maximum parallelism must fulfill the following conditions:
+The max parallelism, set on a per-job and per-operator granularity, determines the maximum parallelism to which a stateful operator can scale.
+There is currently **no way to change** the maximum parallelism of an operator after a job has started without discarding that operators state.
```
Comment (fhueske) [inline docs/ops/production_ready.md]: un-highlight "no way to change"? The other aspects have similar issues and are not highlighted.

File: docs/ops/production_ready.md [hunk 4]
```diff
@@ -22,79 +22,54 @@ specific language governing permissions and limitations
...
-stream processing. However, RocksDB can have worse performance than, for example, the memory-based state backends. If
-you are sure that your state will never exceed main memory and blocking the stream processing to write it is not an issue,
-you **could consider** to not use the RocksDB backends. However, at this point, we **strongly recommend** using RocksDB
-for production.
+In general, we recommend avoiding `MemoryStateBackend` in production because it stores its snapshots inside the JobManager as opposed to persistent disk.
+When deciding between `FsStateBackend` and `RocksDB`, it is a choice between performance and scalability.
+`FsStateBackend` is very fast as each state access and update operates on objects on the Java heap; however, state size is limited by available memory within the cluster.
+On the other hand, `RocksDB` can scale based on available disk space and is the only state backend to support incremental snapshots.
+However, each state access and update requires (de-)serialization and potentially reading from disk which leads to average performance that is an order of magnitude slower than the memory state backends.
```
[the remark this answers is not in the document]
Comment (sjwiesman) [reply docs/ops/production_ready.md]: I think it should be clear from the context but this sentence could probably be worded better :)

**Pull request closed without merging**